### DISCUSSION TITLE
Advanced super-stream usage

### DISCUSSION THREAD

**Discussion by jamiechapmanbrn** [rabbitmq/rabbitmq-stream-java-client#292] [Ideas]
I'm working on creating a superstream consumer for apache druid, in order to avoid the challenges of maintaining a kafka cluster and a rabbitmq cluster. Theoretically, there isn't any gaps I'm aware of in the functionality of superstreams themselves in order to plumb this together properly. Practically, there's a few challenges I've run into and observations from the existing druid indexing code for both Amazon Kinesis and Apache Kafka that I think are missing from this client (as well as the  c# client as far as I can tell).

In other streaming clients, it is possible to quite easily find the partitions or shards of any given stream. This is useful for a worker system that needs information in order to batch the jobs to do some kind of work. For druid, these indexing tasks are both CPU and memory intensive as it crunches the data down into an index format. It decides based on the partitions and some configuration information how it should batch the work and allocates some workers as appropriate. In order to split the work correctly, it uses the list of partitions in order to create workers with specific tasks. The equivalent for rabbitmq would be to find the sub-streams of a super-stream. Right now the workaround would be to list the streams and filter them based on string matches.

The client also doesn't really have the ability to decide in any real way what partitions are being read. This means that if I want to dispatch something to read from some set of the sub-streams, I would need to use the streams client, then piece them together and hope I haven't made some ordering mistake.

The last feature that would be helpful here would be a non-subthread model for reading from streams. In the Kafka client, the Druid indexer reads from the stream for a period of time. In the Kinesis client, it reads a set number of messages. These features are useful because Druid works in 'slices'. It reads for a certain amount of time, then breaks that into an indexible chunk it can put into a backing store. Using the existing rabbit streaming consumer, it would involve starting and stopping the client arbitrarily with some kind of timer or hold condition. It's certainly possible, but not ideal. Where that gets more painful, is combined with the fact that a druid indexing task will read from several streams at once, meaning that each stream would start and stop reading at different times.

This is just my perspective using the super streams feature, hopefully this is helpful for guiding development on this tool.

**Comment by acogoluegnes** [comment]
You can use the "low-level" [Client](https://rabbitmq.github.io/rabbitmq-stream-java-client/stable/api/com/rabbitmq/stream/impl/Client.html) API to [look up the partitions](https://rabbitmq.github.io/rabbitmq-stream-java-client/stable/api/com/rabbitmq/stream/impl/Client.html#partitions-java.lang.String-) of a super stream. This API is a Java-to-stream-protocol mapping, it's technically public but not part of the public API, so it can change, but it's acceptable to use in your use case I'd say. At least it can be good enough to explore the feasibility of a use case.

I'm not sure I understand the use case of the second request, that is just reading a part of the partitions. How would you decide which partitions should be read? With some sort of filter on the partition name like a wildcard? If this is what you expect, it should be easy enough to implement, but it may change the single active consumer semantics quite significantly, we would need to take this into account.

For the last request - consuming messages and stopping automatically IIUC - it could mean having some "condition" API that gets evaluated on the message or chunk arrival and that would trigger the closing of the subscription. What could be the condition? Number of messages, time?

The super stream API for this client library has been designed to be almost invisible to applications, and I think you're hitting the other side of this design because you need insight into the super stream topology.

These are reasonable requests but I don't know if they could have some value for other users. Nevertheless, we can keep refining the semantics and see how we can implement them.

Last thing:

> I'm working on creating a superstream consumer for apache druid, in order to avoid the challenges of maintaining a kafka cluster and a rabbitmq cluster.

I think you meant "Amazon Kinesis" or "Zookeeper" cluster, not "RabbitMQ" cluster, because you'll need a RabbitMQ cluster anyway.

**Reply by jamiechapmanbrn** [reply]
Looking at the client API, I see it has a partitions call, that's great!

I'm not sure I fully understand the use case for the second outside my own, but the Kafka client and Kinesis client both take an iterator or list to determine the specific partitions to be read. As for how it decides, I haven't read the code as it's abstracted away from the readers itself, but presumably the scheduler takes the set of partitions and splits it as evenly as it can into these separate jobs. The scheduler checks what partitions are available and the offsets of those before dispatching jobs to read messages from the queues.

A condition API would work for making the API to read until either condition was hit.

I did mean Kafka cluster, as using Kafka in our architecture and moving messages from RabbitMQ to Kafka would be the alternative to using rabbitmq streams directly here.

**Reply by acogoluegnes** [reply]
I'm still not clear about the second request, being able a subset of the super stream partitions. Why do you want to read only a subset of the partitions in the first place and how do you tell the Kafka/Kinesis client libraries to read only from this subset? You mention it's abstracted away, but there must be something in the library that allows you to consume only from some partitions, not from all.

I can give a try to the condition API. The main difficulty is closing properly, especially WRT to offset tracking. Any idea of the kind of conditions that would trigger closing the subscription (number of messages, time, something else)?

**Reply by acogoluegnes** [reply]
And to make sure we're on the same page, have you read the blog articles on [single active consumer](https://blog.rabbitmq.com/posts/2022/07/rabbitmq-3-11-feature-preview-single-active-consumer-for-streams/) and [super streams](https://blog.rabbitmq.com/posts/2022/07/rabbitmq-3-11-feature-preview-super-streams/)?

**Reply by jamiechapmanbrn** [reply]
I have read the blog article on single active consumers and super streams. The difference is I'm working inside the Druid API, and am limited by the streaming interfaces it requires. It explicitly queries for partitions and offsets, then assigns the partitions and offsets to workers and sends them to work. I'm sure that they are using this architectural for a reason rather than simply using the native balancing that both Kafka as well as Kinesis provide, but I don't know the details. It's possible that there's better way for Druid to use the rabbitmq stream client, but that probably involves changes in core Druid.

I'll share some excerpts from the Kafka client so you can see what interfaces it's using here.

```
  @Override
  public void assign(Set<StreamPartition<Integer>> streamPartitions)
  {
    wrapExceptions(() -> consumer.assign(streamPartitions
                                             .stream()
                                             .map(x -> new TopicPartition(x.getStream(), x.getPartitionId()))
                                             .collect(Collectors.toSet())));
  }

  @Override
  public void seek(StreamPartition<Integer> partition, Long sequenceNumber)
  {
    wrapExceptions(() -> consumer.seek(
        new TopicPartition(partition.getStream(), partition.getPartitionId()),
        sequenceNumber
    ));
  }

  @Override
  public void seekToEarliest(Set<StreamPartition<Integer>> partitions)
  {
    wrapExceptions(() -> consumer.seekToBeginning(partitions
                                                      .stream()
                                                      .map(e -> new TopicPartition(e.getStream(), e.getPartitionId()))
                                                      .collect(Collectors.toList())));
  }

  @Override
  public void seekToLatest(Set<StreamPartition<Integer>> partitions)
  {
    wrapExceptions(() -> consumer.seekToEnd(partitions
                                                .stream()
                                                .map(e -> new TopicPartition(e.getStream(), e.getPartitionId()))
                                                .collect(Collectors.toList())));
  }
  
    @Nonnull
  @Override
  public List<OrderedPartitionableRecord<Integer, Long, KafkaRecordEntity>> poll(long timeout)
  {
    List<OrderedPartitionableRecord<Integer, Long, KafkaRecordEntity>> polledRecords = new ArrayList<>();
    for (ConsumerRecord<byte[], byte[]> record : consumer.poll(Duration.ofMillis(timeout))) {

      polledRecords.add(new OrderedPartitionableRecord<>(
          record.topic(),
          record.partition(),
          record.offset(),
          record.value() == null ? null : ImmutableList.of(new KafkaRecordEntity(record))
      ));
    }
    return polledRecords;
  }
  ```

**Reply by acogoluegnes** [reply]
OK, thanks for the code excerpts, I'll have a look.

**Reply by jamiechapmanbrn** [reply]
I've noticed that the it would benefit from the ability to get the latest and earliest sequence number, it has an option in the config to start from now rather than the beginning of the stream. I see the queryStreamStats function returns the first available offset, but not the latest. I can get the latest by listening to the stream at latest and grabbing the first message and passing back the id, but it's not quite as nice. Poll would improve that considerably though.

**Reply by acogoluegnes** [reply]
I had the look at the [record supplier Druid API](https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/common/RecordSupplier.java) and the [Kafka implementation](https://github.com/apache/druid/blob/master/extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaRecordSupplier.java). They are quite close to each other, I assume the abstraction is based on the implementation.

The API expects a seekable data stream with explicit references to partitions and RabbitMQ super streams are none of this TBH, so expect the implementation to be more involved than the Kafka one because of these discrepancies. RabbitMQ streams use a push-based approach (as opposed to pull-based for Kafka). Super streams are a partitioning solution to help scale out and the stream Java client strives to make it a transparent implementation detail. This is different in the Druid API and in Kafka in general where partitions are fully part of the client API.

I had a quick look at the [Kinesis implementation](https://github.com/apache/druid/blob/master/extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java) and it seems to be closer to what an implementation for RabbitMQ streams could be.

Here are actionable takeaways:
* you can still use super streams for the topology but you may have to use individual consumers because Druid assigns partitions to the record supplier. Or you may use the composite consumer implementation for super streams, but you may have to do some filtering as the Kinesis implementation does.
* the implementation will have to maintain more state than the Kafka one.
* `queryStreamStats` seem to return the information you need, the first offset and the committed chunk ID, which is a good approximation of the last offset in the stream. See the [Javadoc](https://github.com/rabbitmq/rabbitmq-stream-java-client/blob/eaa3d911956dd3fa113d1d14d24e0f6767fbf533/src/main/java/com/rabbitmq/stream/StreamStats.java#L25-L47).

If you have some code to share I can have a look at it.

**Reply by jamiechapmanbrn** [reply]
Sorry for the delay, I got sidetracked by a few other tasks that have come up.
I have a bare bones implementation available here:
https://github.com/apache/druid/pull/14137

I could use suggestions on a more permanent algorithm to prevent opening/closing consumers as often.

Thanks so much for your help!

**Reply by acogoluegnes** [reply]
Thanks for the follow-up. I had a look at the code, I saw you chose to maintain an individual consumer for each super stream partition, which I guess is fine for this use case.

Note streams and super streams do not support "exactly-once delivery" as you state in the PR description. They support deduplication on the publishing side (no duplicated messages in the stream if used properly) but a consumer can see the same message under some circumstances (e.g. when restarting after a failure). Exactly-once delivery between a resource (e.g. a stream) and another (e.g. Druid index) is not possible unless the 2 resources synchronize themselves with a transaction mechanism like XA.

About opening and closing consumers, it's hard to say without more information on the usage: frequency of `poll`, request offset on a given partition (are they contiguous between `poll` calls). The best way is to experiment with real-world scenarios.
